securedrop_client/queue.py

```diff
@@ -68,15 +73,32 @@ def __init__(self, api_client: API, session_maker: scoped_session) -> None:
         self.main_thread.started.connect(self.main_queue.process)
         self.download_file_thread.started.connect(self.download_file_queue.process)
 
-    def start_queues(self, api_client: API) -> None:
+    def logout(self) -> None:
+        self.main_queue.api_client = None
+        self.download_file_queue.api_client = None
+
+    def login(self, api_client: API) -> None:
+        logger.debug('Passing API token to queues')
         self.main_queue.api_client = api_client
         self.download_file_queue.api_client = api_client
+        self.start_queues()
 
-        self.main_thread.start()
-        self.download_file_thread.start()
+    def start_queues(self) -> None:
+        if not self.main_thread.isRunning():
+            logger.debug('Starting main thread')
+            self.main_thread.start()
+
+        if not self.download_file_thread.isRunning():
+            logger.debug('Starting download thread')
+            self.download_file_thread.start()
 
     def enqueue(self, job: ApiJob) -> None:
+        # First check the queues are started in case they died for some reason.
+        self.start_queues()
```

Comment:
[Nit] This makes it so both login and enqueue can start the queue threads, which will create the signal to begin the process loop to start processing jobs. We'll have to remember to always first check that a user is logged in before enqueuing a job then. It almost makes the login function seem unnecessary. It seems like we should add a new login check if the queues have died for some reason.

Reply:
hmm i see what you're saying here: I put `start_queues()` here because I figured any situation where we enqueue jobs is one where we expect the queues to be running, so I thought that was a good place to start them up if they failed for some reason... I can't think of a better place to put this logic (with respect to issue #380). that said to your point, it would probably be worthwhile defense in depth to ensure that we bail and don't enqueue the job if not authenticated (with respect to 1), so added in 86375ff44b2a72fc432d810e2dec3cfe39a1d6cd

as background my thoughts regarding login/logout were:  `login` is the one place that gets the SDK API object and passes it to the queues, and `logout` revokes it from the queues. `start_queues` just makes sure that the queues start up only if they weren't already running - using the existing auth state that the queues have access to. the queue auth state is preserved if the threads are terminated, i.e. `start_queues` can't authenticate the queues, only `login` can do that.

and you are definitely right that we should:

1. Not be adding jobs to the queue if we're not logged in (why I added the `if self.api` block in the `on_submission_download` method on the controller) - while they won't run due to 2, adding jobs when not authed should imho not be done for defense in depth: if we weren't adding jobs to the queue when not logged in, then we would not have been able to download files after logout, even though bug #397 was in place. 
2. Jobs left in the queue after logout should fail (implemented in this PR). when we make the queue persistent we would want to modify this to "pause jobs left in the queue after logout", but until we have persistence this should be the behavior. 
3. And of course we should implement #223 - bearing in mind that this like all server actions can also fail so we still need to carefully manage auth state client-side

Reply:
Agreed, if a user is not logged in, we should not add jobs. Defense in depth looks good.